cpu/esp: improve thread safety in newlib locking functions

gschorcht opened the pull request:
### Contribution description

This PR improves the thread safety of the newlib locking functions as following:

- Dynamic allocation and initialization of the mutex used by a newlib locking variable must not be interrupted. Since a thread context switch can occur on exit from an ISR, the allocation and initialization of such a mutex is guarded by disabling interrupts. The same is done for the release of such a mutex.
- Locking functions implicitly allocate and initialize the mutex used by a newlib locking variable if necessary. This implicit allocation and initialization as well as the subsequent acquiring must not be interrupted. Since a thread context switch can occur on exit from an ISR, the implicit allocation and initialization and the subsequent acquiring of such a newlib locking variable must be guarded by disabling interrupts. The same must be done for the release of such a newlib locking variable.
- Module `malloc_thread_safe` is not needed any longer.

This PR fixes issue #18534 and reverts the change made in PR #18535.

### Testing procedure

```
BOARD=esp32-wroom-32 make flash test -C tests/thread_priority_inversion flash test
```
should work with this PR.
```
BOARD=esp32-wroom-32 make-C tests/malloc_thread_safe flash test 
```
should still work.

### Issues/PRs references

Fixes #18534

maribu commented on cpu/esp_common/syscalls.c:
```diff
@@ -82,13 +82,16 @@ void IRAM_ATTR _lock_init(_lock_t *lock)
     }
     /* cppcheck-suppress memleak; mtx is stored in lock */
+
+    irq_restore(state);
```
I think it would be safe to move this up directly after the call to `malloc()`

gschorcht replied:
Hm, as long as `mtx` is not assigned `*lock`, the locking variable looks to be uninitialized. That is, another thread could try to reinitialize the same variable if a context switch occurs after `malloc` what might be the case if interrupts are pending and the interrupts are enabled. This wouldn't be a problem if all locking variables were initialized before scheduling becomes active. Unfortunately, most locking variables in the newlib are implicitly initialized when `_lock_acquire_*` is called the first time. That's why I guarded the whole function.

maribu commented on cpu/esp_common/syscalls.c:
```diff
@@ -102,13 +105,16 @@ void IRAM_ATTR _lock_init_recursive(_lock_t *lock)
     }
     /* cppcheck-suppress memleak; rmtx is stored in lock */
+
+    irq_restore(state);
```
Same

maribu commented on cpu/esp_common/syscalls.c:
```diff
@@ -142,13 +156,17 @@ void IRAM_ATTR _lock_acquire(_lock_t *lock)
 
     assert(lock != NULL);
 
+    uint32_t state = irq_disable();
```
This should not be needed, the `mutex_lock()` below should disable IRQs at critical sections already

gschorcht replied:
This is what I thought too. But without disabling interrupts, `tests/malloc_thread_safe` crashes on ESP32.

maribu commented on cpu/esp_common/syscalls.c:
```diff
@@ -176,13 +194,17 @@ void IRAM_ATTR _lock_acquire_recursive(_lock_t *lock)
 
     assert(lock != NULL);
 
+    uint32_t state = irq_disable();
```
Same

gschorcht commented:
@maribu Two questions regarding this PR:
- The stack size `THREAD_STACKSIZE_SMALL` for the duelling threads is too small for ESP8266 in `tests/malloc_thread_safe`. What would be the best way from your point of view to increase it only for ESP8266. Should I override `THREAD_STACKSIZE_SMALL` in `cpu/esp8266/include/cpu_conf.h`?
- What might be the reason that the following assertion fails https://github.com/RIOT-OS/RIOT/blob/b024458b432410e232278cc889002b31d0ec53ea/core/lib/rmutex.c#L131 if interrupts are disabled in `_lock_acquire_recursive` here https://github.com/RIOT-OS/RIOT/blob/432a57b390d94985aa95c5e8e22ddeadd80602fe/cpu/esp_common/syscalls.c#L197?

maribu commented on cpu/esp_common/syscalls.c:
```diff
@@ -194,16 +216,22 @@ int IRAM_ATTR _lock_try_acquire(_lock_t *lock)
 
     assert(lock != NULL);
 
+    uint32_t state = irq_disable();
```
Same

maribu commented on cpu/esp_common/syscalls.c:
```diff
@@ -215,16 +243,22 @@ int IRAM_ATTR _lock_try_acquire_recursive(_lock_t *lock)
 
     assert(lock != NULL);
 
+    uint32_t state = irq_disable();
```
Same

maribu reviewed (commented):
Code looks good to me. If I didn't miss anything, no IRQ disabling is needed in the lock acquire functions.

maribu commented:
> The stack size THREAD_STACKSIZE_SMALL for the duelling threads in tests/malloc_thread_safe is too small for ESP8266. What would be the best way from your point of view to increase it only for ESP8266. Should I override THREAD_STACKSIZE_SMALL in cpu/esp8266/include/cpu_conf.h?

Sounds good to me. Those definitions are a bit finicky, but to my understanding the basic idea is that if a given program runs fine on Cortex M with the given stack size define, it *should* also run on all others.

> What might be the reason that the following assertion fails for ESP8266 [...]

The assertion should actually detect when one thread is unlocking the rmutex owned by another (something that is perfectly fine for mutex, btw). Assuming that lock operations and unlock operations are matching (that is the same number of unlocks as locks is issued and only an rmutex actually owned by the calling thread is unlocked), I would guess some kind of memory corruption has happened.

gschorcht commented:
I had to change a lot to get the locking working for both ESP8266 and ESP32:

- The initialization of locking variables with retargetable locking had to be added to avoid their implicit initialization in `_lock_acquire*` functions. Now the locking variables are either statically defined and initialized or they have to be allocated and initialized explicitly by calling `_lock_init*`. `assert`s ensure that locking variables are initialized when `_lock_acquire*` functions are called.
- For ESP32, the operations on recursive locking variables had to be guarded by disabling interrupts to prevent unintended context switches. For ESP8266, interrupts must not be disabled, otherwise the intended context switch doesn't work when trying to lock a `rmutex` that is already locked by another thread. That was the reason for the `assert` in `rmutex_unlock`.

With these changes `tests/malloc_thread_safety` and `tests/thread_priority_inversion` work absolutely stable on ESP32. Without these changes, `tests/malloc_thread_safety` crashes often, probably because of unintended context switches on exit from ISRs. I wasn't able to catch the problem in debugger.

gschorcht commented:
Compilation was successful in CI, however 225 of 227 tests failed for ESP32 [log](https://ci.riot-os.org/RIOT-OS/RIOT/18544/670cfea9943f23bfe052cdb5b047b7b5a44fd544/output.html) with error message:
```
make: *** No rule to make target 'flash-only'.  Stop.
```
which seems to be completely unrelated.

MrKevinWeiss commented:
Can you limit your build so we can test a bit more?

```diff
diff --git a/.murdock b/.murdock
index e2432b2c09..49f1f98b96 100755
--- a/.murdock
+++ b/.murdock
@@ -1,9 +1,9 @@
 #!/bin/sh
 
 # uncomment and change this to limit builds, e.g.,
-#export BOARDS="samr21-xpro native"
+export BOARDS="esp32-wroom-32"
 # and / or
-#export APPS="examples/hello-world tests/unittests"
+export APPS="tests/thread_priority_inversion"
 
 #
 : ${TEST_BOARDS_AVAILABLE:="esp32-wroom-32 samr21-xpro"}
```

MrKevinWeiss commented:
AFAIK the pi fleet only has 1 esp32-wroom-32 still working, if that has gone down than hardware tests may be a bit of a challenge.

MrKevinWeiss commented:
Whoa, the HiL actually did what I wanted it too.  The pi-fleet is in fact broken now.  I think @kaspar030 is on it.  Meanwhile it looks like this PR fixes the one test.  It seems like there is a problem in the `arduino-mega2560` unrelated to this PR but probably good to look into.

gschorcht commented:
> * The initialization of locking variables with retargetable locking had to be added to avoid their implicit initialization in `_lock_acquire*` functions. Now the locking variables are either statically defined and initialized or they have to be allocated and initialized explicitly by calling `_lock_init*`. `assert`s ensure that locking variables are initialized when `_lock_acquire*` functions are called.

After some further testing, I'm afraid I have to admit that it's not possible after all to assume that a lock variable is already assigned and initialized the first time `_lock_acquire*` is used. When `esp_wifi` is used, some ESP IDF sources are used that use various static lock variables without allocating and initializing them using `_lock_init*`. They simply use `_lock_acquire*` and `_lock_release` and rely on implicit allocation and initialization.

That is, I have to revert some changes. @maribu Would it perhaps be better to completely squash and rebase the changes rather than fix the fixes of the fixes?

maribu commented:
Do as you consider is best.

And more generally: My review-workflow doesn't really profit from the "squash only when asked" policy. So if a PR has only reviews of me, you can safely assume implicit approval to squashing at will ;)

maribu commented:
Oh, regarding squashing: @MrKevinWeiss also reviewed this PR. Sorry that I lost track. This omission was not intentional, but human error due to more context switching between task than my brain can safely handle :/

Anyway, from my side you have both implicit and explicit support for squashing ;)

gschorcht commented:
> Can you limit your build so we can test a bit more?

@MrKevinWeiss [Log](https://ci.riot-os.org/RIOT-OS/RIOT/18544/5b713f7b621ba23974179f33c7ef15961c8e653f/output.html)

gschorcht commented:
@maribu With commits be0e5307b2c95875253cfe5690ace0035ceb0478 and 2b1ab9f84635b65b380dc0e2823fedd48b01f85d `tests/thread_priority_inversion` also works for ESP8266. Should I split off these commits as a separate PR or would the change of the title to cover all the commits ok?

MrKevinWeiss commented:
Just do it in the one pr. I can start another hil test when I'm free to see if the esp8266 is fixed too.

gschorcht commented:
All tests runs were now successful [log](https://ci.riot-os.org/RIOT-OS/RIOT/18544/966930e0b087c7bd1df5c8a32252f3c183a7ce89/output.html).

maribu reviewed (commented):
Code looks good to me. The CI found a few typos and one instance of two subsequent blank lines.

The doccheck will need and entry in `exclude_pattern` to not complain about the stack size (or an Doxygen comment). Both is fine with me.

gschorcht commented:
> Code looks good to me. The CI found a few typos and one instance of two subsequent blank lines.
> 
> The doccheck will need and entry in `exclude_pattern` to not complain about the stack size (or an Doxygen comment). Both is fine with me.

I squashed the fixes of static tests directly to avoid additional CI load by an extra compilation test cycle. Since all tests were already successful ([log](https://ci.riot-os.org/RIOT-OS/RIOT/18544/966930e0b087c7bd1df5c8a32252f3c183a7ce89/output.html)), I removed label `CI: run tests` for these "cosmetic" corrections.

maribu reviewed (approved):
ACK. Looks good to me and the CI did the testing :)
